### DISCUSSION TITLE
variable headers and footers in my e107 themes

### DISCUSSION THREAD

**Discussion by Jimmi08** [e107inc/e107#5226] [Ideas]
This is the way how I do this stuff.  The only problem is not working Menu manager (reported).    
It is pretty easy to maintain, maybe this helps somebody too.

Very simple **theme.html**

```
<body class="{LAYOUT_ID} {WP_BODY_TAG}" {BODY_ONLOAD}>
		{---HEADER---}
		{---LAYOUT---}
		{SETSTYLE=footer}
		{---FOOTER---}
		{---MODAL---}
</body> 
```
  
Structure of HTML headers and footers
![image](https://github.com/e107inc/e107/assets/5429548/06ce4b99-a69b-42f4-9536-50ff828e2cb8)

**Theme shortcode:**
```
    /**
     * Special Header Shortcode for dynamic menuarea templates.
     * @shortcode {---HEADER---}
     * @return string
     */
    function sc_header()
    {
        if (defined('THEME_LAYOUT'))
        {
            switch (THEME_LAYOUT)
            {
                case "splash":
                case "error":
                    return "";
                    break;
                case "home":
                    $headerName = "header_home.html";
                    break;
                case "userpanel":
                    $headerName = "header_userpanel.html";
                    break;    
                default:
                    $headerName = "header_home.html";
                    break;
            }
            $theme = e107::getPref('sitetheme');
            $headerpath = e_THEME . $theme . "/headers/" . $headerName;
            if (file_exists($headerpath))
            {
                $text = file_get_contents($headerpath);
            }
            return $text;
        }
        return "";
    }
```

See? No iframe is needed. If I don't need header, it is set by layout. 

The footer is done similar way:

```
/**
     * Special Header Shortcode for dynamic menuarea templates.
     * @shortcode {---FOOTER---}
     * @return string
     */
    function sc_footer()
    {
        if (defined('THEME_LAYOUT'))
        {
            switch (THEME_LAYOUT)
            {
                case "splash":
                case "error":
                    return "";
                    break;
                case "home":
                    $headerName = "footer_home.html";
                    break;
                case "userpanel":
                    $headerName = "footer_userpanel.html";
                    break;
                default:
                    $headerName = "footer_home.html";
                    break;
            }
            $theme = e107::getPref('sitetheme');
            $headerpath = e_THEME . $theme . "/footers/" . $headerName;
            if (file_exists($headerpath))
            {
                $text = file_get_contents($headerpath);
            }
            return $text;
        }
        return "";
    }
```

Menu manager issue:
It is not so big problem with CMENU and NAVIGATION shortcodes.  Or I will find some other solution.  

![image](https://github.com/e107inc/e107/assets/5429548/84577ef1-257e-47e2-844d-a8c2792cfba3)

**Comment by Jimmi08** [comment]
Still problem.  The logic for creating list of available menus is different for left and right side of menu manager  Right side is correct.

![image](https://github.com/e107inc/e107/assets/5429548/a0b93c14-b747-4b0c-b306-a83b1481d737)

**Reply by tgtje** [reply]
@Jimmi08 do not forget (using a earlier posted image) : you do have a sidebar left layout ?

**Reply by rica-carv** [reply]
Sorry to "bump in", but i don't know if it's related or not, but i have a very weird strange situation in my theme, when using the "challeging" menu management admin option.
If i click in a menu view options icon, i end up with this:
![image](https://github.com/e107inc/e107/assets/17130530/21b2e96d-4b85-4bd9-875a-6257e03d6dc7)

By the way, my theme is the old fizi bs4 news theme, updated to BS5 and with a lot of customization.

What am i doing wrong here to be unable to manage menus?
Do i have to tamper the database to manage menus in my theme?

**Reply by Jimmi08** [reply]
Oh, this so familiar... 2-3 years ago. Check the console (on both frames - menu manager and frontend theme itself)
- it was something with wrong parsing, the front theme influenced the menu manager. 
I tried to find that issue but I wasn't successful.

**Reply by rica-carv** [reply]
> Oh, this so familiar... 2-3 years ago. Check the console (on both frames - menu manager and frontend theme itself)
> 
>     * it was something with wrong parsing, the front theme influenced the menu manager.
>       I tried to find that issue but I wasn't successful.

@Jimmi08 Just catched that little pesky bug. It was the tooltips js. Just removed the line `e107::js("footer-inline", 	"$('.e-tip').tooltip({container: 'body'})"); // activate bootstrap tooltips. ` and i got the menus visibility definition back...

![image](https://github.com/e107inc/e107/assets/17130530/f00b6a61-4e4b-4609-85ba-7bb897081410)

Guess i got to find another compatible tooltip js code.
Google, here i go.....

**Reply by rica-carv** [reply]
@Jimmi08 Solved the visibility menu management, but still got the same issue as you, regarding the assignement of menus to areas of the theme.
Currently i'm assigning menus via database...
I'm wondering if the issue isn't Bootstrap 5 related, since i'm using bootstrap 5.3l3 (latest released), and i'm force loading it in theme.php, since it's not available in e107 core...

**Reply by tgtje** [reply]
@rica-carv  bs 5 issue ??  hmm always thought theres something happening with/in , allthough i more suspect the newer jQuery-3.7.1.
More a 'feeling' than actual blaming it.. (can't proof, in regards when using php 7.x vs higher 8).

**Reply by Jimmi08** [reply]
Yes, I have a strong feeling that something is wrong, just in some cases...  I created many issues related to this.  I found the reason for many of them.  And the result?  I gave up creating of new themes.  

`Currently i'm assigning menus via database...`  - I had plugin for this...  I really don't understand why to have so complicated and not working menu manager when you can do the same thing just with editing records via new Admin UI.  

@rica-carv you are running into the same issues I run 2-3 years ago, almost the same - constants, menu manager, jQuery, news...     wait when you get to users stuff.   And you don't use SEF-URLs - it would be the new area of issues.   Do you have any reaction from core developers to your issues/questions? I don't think so, I would get some notifications.  

Good luck.  There is always some  workaround, this is reason why I love e107 so much.

**Reply by rica-carv** [reply]
> almost the same - constants, menu manager, jQuery, news... wait when you get to users stuff

@Jimmi08 news issues, i've been / i'm there

> And you don't use SEF-URLs - it would be the new area of issues.

@Jimmi08 Yes i do or at least TRY to use SEF's. Until now, some minor issues...

> There is always some workaround, this is reason why I love e107 so much.

@Jimmi08 Indeed, you can always workaround ALMOST asny issue. For instance, my theme looks like a mix of E107 band aids 😆

**Comment by Jimmi08** [comment]
Fixed somehow during time...  it works now.
